File: base/script/weapon_shotgun.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____StateMachine = require("StateMachine")
local StateMachine = ____StateMachine.StateMachine
local AnimChannel = AnimChannel or ({})
AnimChannel.All = 0
AnimChannel[AnimChannel.All] = "All"
AnimChannel.Torso = 1
AnimChannel[AnimChannel.Torso] = "Torso"
AnimChannel.Legs = 2
AnimChannel[AnimChannel.Legs] = "Legs"
AnimChannel.Head = 3
AnimChannel[AnimChannel.Head] = "Head"
AnimChannel.Eyelids = 4
AnimChannel[AnimChannel.Eyelids] = "Eyelids"
local SHOTGUN_FIRERATE = 1.333
local SHOTGUN_LOWAMMO = 2
local SHOTGUN_RELOADRATE = 2
local SHOTGUN_NUMPROJECTILES = 13
local SHOTGUN_IDLE_TO_IDLE = 0
local SHOTGUN_IDLE_TO_LOWER = 4
local SHOTGUN_IDLE_TO_FIRE = 1
local SHOTGUN_IDLE_TO_RELOAD = 4
local SHOTGUN_IDLE_TO_NOAMMO = 4
local SHOTGUN_NOAMMO_TO_RELOAD = 4
local SHOTGUN_NOAMMO_TO_IDLE = 4
local SHOTGUN_RAISE_TO_IDLE = 1
local SHOTGUN_FIRE_TO_IDLE = 4
local SHOTGUN_RELOAD_TO_IDLE = 4
local SHOTGUN_RELOAD_TO_FIRE = 4
local SHOTGUN_RELOAD_TO_LOWER = 2
local SND_CHANNEL_ANY = 0
local SND_CHANNEL_VOICE = 1
local SND_CHANNEL_VOICE2 = 2
local SND_CHANNEL_BODY = 3
local SND_CHANNEL_BODY2 = 4
local SND_CHANNEL_BODY3 = 5
local SND_CHANNEL_WEAPON = 6
local SND_CHANNEL_ITEM = 7
local SND_CHANNEL_HEART = 8
local SND_CHANNEL_PDA = 9
local SND_CHANNEL_DEMONIC = 10
local Shotgun = __TS__Class()
Shotgun.name = "Shotgun"
function Shotgun.prototype.____constructor(self)
end
function Shotgun.prototype.init(self)
    local stateMachine = __TS__New(StateMachine)
    local weapon = self
    stateMachine:addState({
        name = "Initial",
        onEnter = function()
            sys:println("Initial state")
            weapon.spread = self:getIntKey("spread")
            self.nextAttack = sys:getTime()
            weapon:weaponState("Raise", 0)
        end
    }).addState({
        name = "Raise",
        onEnter = function(____, task)
            sys:println("Raise state")
            weapon:weaponRising()
            weapon:playAnim(AnimChannel.All, "raise")
            task:create(function()
                while weapon:animDone(AnimChannel.All, SHOTGUN_RAISE_TO_IDLE) ~= 1 do
                    task:wait(0)
                end
                weapon:weaponState("Idle", 1)
            end)
        end
    }).addState({
        name = "Lower",
        onEnter = function(____, task)
            weapon:weaponLowering()
            weapon:playAnim(AnimChannel.All, "putaway")
            task:create(function()
                while weapon:animDone(AnimChannel.All, SHOTGUN_RAISE_TO_IDLE) ~= 1 do
                    task:wait(0)
                end
                weapon:weaponHolstered()
                weapon:weaponState("Idle", 1)
            end)
        end,
        onRaiseWeapon = function()
            weapon:setWeaponState("Raise")
        end
    }).addState({
        name = "Idle",
        onEnter = function(____, task)
            sys:println("Idle state")
            if weapon:ammoInClip() == 0 then
                sys:println("weapon out of ammo")
                weapon:weaponOutOfAmmo()
            else
                sys:println("weapon ready")
                weapon:weaponReady()
            end
            weapon:playCycle(AnimChannel.All, "idle")
        end,
        onLowerWeapon = function()
            weapon:weaponState("Lower", SHOTGUN_IDLE_TO_LOWER)
        end,
        onBeginAttack = function()
            sys:println("Attack")
            local currentTime = sys:getTime()
            local ammoClip = weapon:ammoInClip()
            local ammoAvailable = weapon:ammoAvailable()
            if currentTime >= weapon.nextAttack then
                if ammoClip > 0 then
                    weapon.nextAttack = sys:getTime() + SHOTGUN_FIRERATE
                    weapon:weaponState("Fire", SHOTGUN_IDLE_TO_FIRE)
                elseif ammoAvailable > 0 then
                    if weapon:autoReload() then
                        weapon:netReload()
                        weapon:weaponState("Reload", SHOTGUN_IDLE_TO_RELOAD)
                    else
                        weapon:weaponState("NoAmmo", SHOTGUN_IDLE_TO_NOAMMO)
                    end
                else
                    weapon:weaponState("NoAmmo", SHOTGUN_IDLE_TO_NOAMMO)
                end
            end
        end,
        onReload = function()
            local ammoClip = weapon:ammoInClip()
            local clipSize = weapon:clipSize()
            local ammoAvailable = weapon:ammoAvailable()
            if ammoAvailable > ammoClip and ammoClip < clipSize then
                weapon:netReload()
                weapon:weaponState("Reload", SHOTGUN_IDLE_TO_RELOAD)
            end
        end,
        onNetReload = function()
            weapon:weaponState("Reload", SHOTGUN_IDLE_TO_RELOAD)
        end
    }).addState({
        name = "Fire",
        onEnter = function(____, task)
            sys:println("Start fire")
            local ammoClip = weapon:ammoInClip()
            if ammoClip == SHOTGUN_LOWAMMO then
                weapon:startSound("snd_lowammo", SND_CHANNEL_ITEM, 1)
            end
            weapon:launchProjectiles(
                SHOTGUN_NUMPROJECTILES,
                weapon.spread,
                0,
                1,
                1
            )
            weapon:playAnim(AnimChannel.All, "fire")
            task:create(function()
                while weapon:animDone(AnimChannel.All, SHOTGUN_FIRE_TO_IDLE) ~= 1 do
                    task:wait(0)
                end
                weapon:weaponState("Idle", 1)
            end)
        end,
        onWeaponNetReload = function()
            weapon:weaponState("Reload", SHOTGUN_IDLE_TO_RELOAD)
        end
    }).addState({
        name = "Reload",
        onEnter = function(____, task)
            weapon.isAttacking = false
            weapon.lowerWeapon = false
            weapon.endReload = false
            sys:println("Start reload")
            local clipSize = weapon:clipSize()
            local ammoAvail = weapon:ammoAvailable()
            local ammoClip = weapon:ammoInClip()
            weapon:weaponReloading()
            weapon:playAnim(AnimChannel.All, "reload_start")
            task:create(function()
                while weapon:animDone(AnimChannel.All, 0) ~= 1 do
                    task:wait(0)
                end
                while ammoClip < clipSize and ammoClip < ammoAvail do
                    weapon:weaponReloading()
                    weapon:playAnim(AnimChannel.All, "reload_loop")
                    weapon:waitFrame()
                    if weapon.isAttacking and ammoClip > 0 then
                        break
                    end
                    if weapon.lowerWeapon or weapon.endReload then
                        break
                    end
                    while weapon:animDone(AnimChannel.All, 0) ~= 1 do
                        task:wait(0)
                    end
                    weapon:addToClip(SHOTGUN_RELOADRATE)
                    weapon:weaponReady()
                    task:wait(0)
                    if weapon.isAttacking or weapon.lowerWeapon or weapon.endReload then
                        break
                    end
                    ammoAvail = weapon:ammoAvailable()
                    ammoClip = weapon:ammoInClip()
                end
                weapon:netEndReload()
                weapon:playAnim(AnimChannel.All, "reload_end")
                task:wait(0)
                while weapon:animDone(AnimChannel.All, SHOTGUN_RELOAD_TO_IDLE) == 0 do
                    if weapon.lowerWeapon then
                        weapon:weaponState("Lower", SHOTGUN_RELOAD_TO_LOWER)
                        return
                    elseif weapon.isAttacking then
                        weapon:weaponState("Fire", SHOTGUN_RELOAD_TO_FIRE)
                        return
                    end
                    task:wait(0)
                end
                weapon:weaponState("Idle", SHOTGUN_RELOAD_TO_IDLE)
            end)
        end,
        onBeginAttack = function()
            weapon.isAttacking = true
        end,
        onEndAttack = function()
            weapon.isAttacking = false
        end,
        onLowerWeapon = function()
            weapon.lowerWeapon = true
        end,
        onEndReload = function()
            weapon.endReload = true
        end
    })
end
function Shotgun.prototype.think(self)
end
return ____exports

File: base/script/WeaponShotgun.ts
import { RegisterStateEntity } from "./EntitySystem"
import { AnimChannel, SoundChannel } from "./Global"
import { StateMachine } from "./StateMachine"
import { Task } from "./Task"
import { DefaultWeaponState, WeaponStateEntity } from "./Weapon"
import "./imgui"

const SHOTGUN_FIRERATE = 1.333
const SHOTGUN_LOWAMMO = 2
const SHOTGUN_RELOADRATE = 2
const SHOTGUN_NUMPROJECTILES = 13

// blend times
const SHOTGUN_IDLE_TO_LOWER = 4
const SHOTGUN_IDLE_TO_FIRE = 1
const SHOTGUN_IDLE_TO_RELOAD = 4
const SHOTGUN_IDLE_TO_NOAMMO = 4
const SHOTGUN_NOAMMO_TO_RELOAD = 4
const SHOTGUN_NOAMMO_TO_IDLE = 4
const SHOTGUN_RAISE_TO_IDLE = 1
const SHOTGUN_FIRE_TO_IDLE = 4
const SHOTGUN_RELOAD_TO_IDLE = 4
const SHOTGUN_RELOAD_TO_FIRE = 4
const SHOTGUN_RELOAD_TO_LOWER = 2

class WeaponShotgun extends WeaponStateEntity {

    spread: number = 0;
    nextAttack: number = 0;
    endReload: boolean = false;
    isAttacking: boolean = false;
    isNetReload: boolean = false;
    isRaiseWeapon: boolean = false;

    constructor(entity: idWeapon) {
        super(entity);
        this.spread = this.entity.getFloatKey("spread");
        sys.println(`Constructed shotgun for ${this.entity.getName()}, id = ${this.entity.getEntityNum()}`);
    }

    stateMachine = new StateMachine<DefaultWeaponState>({
        initial: "Initial",
        states: {
            Initial: {
                onEnter: (): void => {
                    sys.println("Initial state")
                    this.nextAttack = sys.getTime()
                    this.entity.weaponState("Raise", 0)
                }
            },

            Raise: {
                onEnter: (task: Task): void => {
                    sys.println("Raise state");
                    this.entity.weaponRising();
                    this.entity.playAnim(AnimChannel.All, "raise");
                    task.create((): void => {
                        while (this.entity.animDone(AnimChannel.All, SHOTGUN_RAISE_TO_IDLE) != 1) {
                            task.waitFrame();
                        }
                        this.entity.weaponState("Idle", 1);
                    });
                }
            },

            Lower: {
                onEnter: (task: Task): void => {
                    sys.println("lowering")
                    this.entity.weaponLowering();
                    this.entity.playAnim(AnimChannel.All, "putaway");
                    task.create((): void => {
                        while (this.entity.animDone(AnimChannel.All, SHOTGUN_RAISE_TO_IDLE) != 1) {
                            task.waitFrame();
                        }
                        this.entity.weaponHolstered();
                    })
                },
                onRaiseWeapon: (): void => {
                    return this.entity.setWeaponState("Raise");
                }
            },

            Idle: {
                onEnter: (task: Task): void => {
                    sys.println("Idle state");
        
                    if (this.entity.ammoInClip() == 0) { 
                        sys.println("weapon out of ammo");
                        this.entity.weaponOutOfAmmo();
                        return;
                    } else {
                        sys.println("weapon ready");
                        this.entity.weaponReady();
                    }

                    task.create((): void => {
                        sys.println("Before loop");
                        while (true) {
                            if (imgui.isReadyToRender()) {
                                imgui.beginWindow("Weapon state", imgui.WindowFlags.NoNavFocus);
                                imgui.text("Idle state");
                                imgui.endWindow();
                            }
                            task.waitFrame();
                        }
                    });
                
                    this.entity.playCycle(AnimChannel.All, "idle");
                },
                onLowerWeapon: (): void => {
                    this.entity.weaponState("Lower", SHOTGUN_IDLE_TO_LOWER);
                    this.isRaiseWeapon = false;
                },
                onHolster: (): void => {
                    this.isRaiseWeapon = false;
                },
                onBeginAttack: (): void => {
                    sys.println("Attack");
        
                    const currentTime = sys.getTime();
                    const ammoClip = this.entity.ammoInClip();
                    const ammoAvailable = this.entity.ammoAvailable();
        
                    if (currentTime >= this.nextAttack) {
                        if (ammoClip > 0) {
                            this.nextAttack = sys.getTime() + SHOTGUN_FIRERATE;
                            this.entity.weaponState("Fire", SHOTGUN_IDLE_TO_FIRE);
                        }
                        else if (ammoAvailable > 0) {
                            if (this.entity.autoReload()) {
                                this.entity.netReload();
                                this.entity.weaponState("Reload", SHOTGUN_IDLE_TO_RELOAD);
                            }
                            else {
                                this.entity.weaponState("NoAmmo", SHOTGUN_IDLE_TO_NOAMMO);
                            }
                        }
                        else {
                            this.entity.weaponState("NoAmmo", SHOTGUN_IDLE_TO_NOAMMO);
                        }
                    }
                },
                onReload: (): void => {
                    const ammoClip = this.entity.ammoInClip();
                    const clipSize = this.entity.clipSize();
                    const ammoAvailable = this.entity.ammoAvailable();
                    if (ammoAvailable > ammoClip && ammoClip < clipSize) {
                        this.entity.netReload();
                        this.entity.weaponState("Reload", SHOTGUN_IDLE_TO_RELOAD);
                    }
                },
                onNetReload: (): void => {
                    this.isNetReload = true;
                    this.entity.weaponState("Reload", SHOTGUN_IDLE_TO_RELOAD);
                }
            },
            Fire: {
                onEnter: (task: Task): void => {
                    sys.println("Start fire");
        
                    if (this.isNetReload) {
                        this.isNetReload = false;
                        this.entity.weaponState("Reload", SHOTGUN_IDLE_TO_RELOAD);
                        return;
                    }
            
                    const ammoClip = this.entity.ammoInClip();
                    if (ammoClip == SHOTGUN_LOWAMMO) {
                        this.entity.startSound("snd_lowammo", SoundChannel.Item, 1);
                    }
        
                    this.entity.launchProjectiles(SHOTGUN_NUMPROJECTILES, this.spread, 0, 1.0, 1.0);
                    this.entity.playAnim(AnimChannel.All, "fire");

                    task.create((): void => {
                        while (this.entity.animDone(AnimChannel.All, SHOTGUN_FIRE_TO_IDLE) != 1) {
                            task.waitFrame();
                        }
                        this.entity.weaponState("Idle", SHOTGUN_FIRE_TO_IDLE);
                    })

                    task.create((): void => {
                        while (true) {
                            if (imgui.isReadyToRender()) {
                                imgui.beginWindow("Poop", imgui.WindowFlags.None);
                                imgui.text("Fire state");
                                imgui.endWindow();
                            }
                            task.waitFrame();
                        }
                    });
                },
                onWeaponNetReload: (): void => {
                    this.entity.weaponState("Reload", SHOTGUN_IDLE_TO_RELOAD);
                }
            },
            Reload: {
                onEnter: (task: Task): void => {
                    this.isAttacking = false;
                    this.endReload = false;
        
                    sys.println("Start reload");
        
                    let clipSize = this.entity.clipSize();
                    let ammoAvail = this.entity.ammoAvailable();
                    let ammoClip = this.entity.ammoInClip();
        
                    this.entity.weaponReloading();
                    this.entity.playAnim(AnimChannel.All, "reload_start");
        
                    task.create((): void => {
                        while (this.entity.animDone(AnimChannel.All, 0) != 1) {
                            task.waitFrame();
                        }
                        
                        while (ammoClip < clipSize && ammoClip < ammoAvail) {
                            this.entity.weaponReloading();
                            this.entity.playAnim(AnimChannel.All, "reload_loop");
                            this.entity.waitFrame();
        
                            // allow the player to shoot or switch weapons since shotgun is so slow reloading
                            if (this.isAttacking && ammoClip > 0) {
                                break;
                            }
                        
                            if (!this.isRaiseWeapon || this.endReload) {
                                break;
                            }
                        
                            while (this.entity.animDone(AnimChannel.All, 0) != 1) {
                                task.waitFrame();
                            }
                        
                            this.entity.addToClip(SHOTGUN_RELOADRATE);
                            this.entity.weaponReady();
                            task.waitFrame();
                        
                            if (this.isAttacking || !this.isRaiseWeapon || this.endReload) {
                                // allow the player to shoot or switch weapons since shotgun is so slow reloading
                                break;
                            }
                        
                            ammoAvail = this.entity.ammoAvailable();
                            ammoClip = this.entity.ammoInClip();
                        }
                        
                        this.entity.netEndReload();
                        this.entity.playAnim(AnimChannel.All, "reload_end");
                        task.waitFrame();
                        
                        while (this.entity.animDone(AnimChannel.All, SHOTGUN_RELOAD_TO_IDLE) == 0) {
                            if (!this.isRaiseWeapon) {
                                this.entity.weaponState("Lower", SHOTGUN_RELOAD_TO_LOWER);
                                return;
                            }
                            else if (this.isAttacking) {
                                this.entity.weaponState("Fire", SHOTGUN_RELOAD_TO_FIRE);
                                return;
                            }
                            task.waitFrame();
                        }
                        
                        this.entity.weaponState("Idle", SHOTGUN_RELOAD_TO_IDLE);
                    })
                },
                onBeginAttack: (): void => {
                    this.isAttacking = true;
                },
                onEndAttack: (): void => {
                    this.isAttacking = false;
                },
                onLowerWeapon: (): void => {
                    this.isRaiseWeapon = false;
                },
                onHolster: (): void => {
                    this.isRaiseWeapon = false;
                },
                onEndReload: (): void => {
                    this.endReload = true;
                }
            },
            NoAmmo: {
                onEnter: (task: Task): void => {
                    if (this.isNetReload) {
                        this.isNetReload = false;
                        this.entity.weaponState( "Reload", SHOTGUN_NOAMMO_TO_RELOAD );
                        return;
                    }
                
                    this.entity.playAnim( AnimChannel.All, "noammo" );
                    task.create((): void => {
                        while (this.entity.animDone(AnimChannel.All, SHOTGUN_NOAMMO_TO_IDLE) != 1) {
                            task.waitFrame()
                        }
                        this.entity.weaponState("Lower", SHOTGUN_NOAMMO_TO_IDLE)
                    })
                }
            }
        }
    });
}

RegisterStateEntity(WeaponShotgun);
